src/mkvgen/src/MkvGenerator.c

```diff
@@ -620,6 +620,168 @@ STATUS mkvgenGenerateTag(PMkvGenerator pMkvGenerator, PBYTE pBuffer, PCHAR tagNa
+            break;
+    }
+    packagedSize += MKV_TAG_NAME_BITS_SIZE + MKV_TAG_STRING_BITS_SIZE + tagNameLen + tagValueLen;
+
+    // Check if we are asked for size only and early return if so
+    CHK(pBuffer != NULL, STATUS_SUCCESS);
+
+    // Preliminary check for the buffer size
+    CHK(*pSize >= packagedSize, STATUS_NOT_ENOUGH_MEMORY);
+
+    // Start with the full buffer
+    bufferSize = *pSize;
+
+    // Check the size and copy the structure first
+    switch(parent) {
+        case MKV_TREE_TAGS:
+            encodedLen = MKV_TAGS_BITS_SIZE;
+            CHK(bufferSize >= encodedLen, STATUS_NOT_ENOUGH_MEMORY);
+            MEMCPY(pCurrentPnt, MKV_TAGS_BITS, encodedLen);
+            bufferSize -= encodedLen;
+            pCurrentPnt += encodedLen;
+            break;
+
+        case MKV_TREE_TAG:
+            encodedLen = MKV_TAG_BITS_SIZE;
+            CHK(bufferSize >= encodedLen, STATUS_NOT_ENOUGH_MEMORY);
+            MEMCPY(pCurrentPnt, MKV_TAG_BITS, encodedLen);
+            bufferSize -= encodedLen;
+            pCurrentPnt += encodedLen;
+            break;
+
+        case MKV_TREE_SIMPLE:
+            encodedLen = MKV_TAG_SIMPLE_BITS_SIZE;
+            CHK(bufferSize >= encodedLen, STATUS_NOT_ENOUGH_MEMORY);
+            MEMCPY(pCurrentPnt, MKV_TAG_SIMPLE_BITS, encodedLen);
+            bufferSize -= encodedLen;
+            pCurrentPnt += encodedLen;
+            break;
+
+        default:
+            break;
+
+    }
+
+    // Copy the tag name
+    encodedLen = MKV_TAG_NAME_BITS_SIZE;
+    CHK(bufferSize >= encodedLen + tagNameLen, STATUS_NOT_ENOUGH_MEMORY);
+    MEMCPY(pCurrentPnt, MKV_TAG_NAME_BITS, encodedLen);
+    bufferSize -= encodedLen;
+    pCurrentPnt += encodedLen;
+
+    // Copy the name
+    MEMCPY(pCurrentPnt, tagName, tagNameLen);
+
+    // Reduce the size of the buffer
+    bufferSize -= tagNameLen;
+    pCurrentPnt += tagNameLen;
+
+    // Copy the tag string value
+    encodedLen = MKV_TAG_STRING_BITS_SIZE;
+    CHK(bufferSize >= encodedLen + tagValueLen, STATUS_NOT_ENOUGH_MEMORY);
+    MEMCPY(pCurrentPnt, MKV_TAG_STRING_BITS, encodedLen);
+    bufferSize -= encodedLen;
+    pCurrentPnt += encodedLen;
+
+    // Copy the value
+    MEMCPY(pCurrentPnt, tagValue, tagValueLen);
+
+    // Reduce the size of the buffer
+    bufferSize -= tagValueLen;
+    pCurrentPnt += tagValueLen;
+
+    // Fix-up the tags element size
+    /*
+     * Options will be:
+     *     TAGS + TAG + SIMPLE
+     *     TAG + SIMPLE
+     *     SIMPLE
+     */
+    switch(parent) {
+        case MKV_TREE_TAGS:
+            totalOffset += MKV_TAGS_ELEMENT_SIZE_OFFSET;
+            encodedElementLength = 0x100000000000000ULL | (UINT64)(packagedSize - (totalOffset + MKV_GENERIC_ELEMENT_OFFSET));
+            PUT_UNALIGNED_BIG_ENDIAN((PINT64)(pStartPnt + totalOffset), encodedElementLength);
+            totalOffset += MKV_GENERIC_ELEMENT_OFFSET; 
+
+        case MKV_TREE_TAG:
+            totalOffset += MKV_GENERIC_ELEMENT_SIZE_OFFSET;

```

Comment:
These case statements do the same thing?

Reply:
I'm stilling add comments, but these case statements are purposefully missing a break. TAG and SIMPLE have the same offsets, so their case statements are quite similar but the missing break distinguishes their functionality.